src/Middleware/ForwardedForMiddleware.php

```diff
@@ -0,0 +1,137 @@
+use Amp\Http\Server\Response;
+use Amp\Socket\CidrMatcher;
+use Amp\Socket\InternetAddress;
+use Amp\Socket\SocketException;
+
+final class ForwardedForMiddleware implements Middleware
+{
+    /** @var list<CidrMatcher> */
+    private readonly array $trustedProxies;
+
+    /** @var LocalCache<bool> */
+    private readonly LocalCache $trustedIps;
+
+    /**
+     * @param array<non-empty-string> $trustedProxies Array of IPv4 or IPv6 addresses with an optional subnet mask.
+     *      e.g., '172.18.0.0/24'
+     */
+    public function __construct(
+        private readonly ForwardedForHeaderType $headerType,
+        array $trustedProxies,
+        int $cacheSize = 1000
+    ) {
+        $this->trustedProxies = \array_map(
+            static fn (string $ip) => new CidrMatcher($ip),
+            \array_values($trustedProxies),
+        );
+
+        $this->trustedIps = new LocalCache($cacheSize);
+    }
+
+    public function handleRequest(Request $request, RequestHandler $requestHandler): Response
+    {
+        $clientAddress = $request->getClient()->getRemoteAddress();
+
+        if ($clientAddress instanceof InternetAddress && $this->isTrustedProxy($clientAddress)) {
+            $request->setAttribute(self::class, match ($this->headerType) {
+                ForwardedForHeaderType::FORWARDED => $this->getForwarded($request),
+                ForwardedForHeaderType::X_FORWARDED_FOR => $this->getForwardedFor($request),
+            });
+        } else {
+            $request->setAttribute(self::class, []);
+        }
+
+        return $requestHandler->handleRequest($request);
+    }
+
+    private function isTrustedProxy(InternetAddress $address): bool
+    {
+        $ip = $address->getAddress();
+        $trusted = $this->trustedIps->get($ip);
+        if ($trusted !== null) {
+            return $trusted;
+        }
+
+        $trusted = false;
+        foreach ($this->trustedProxies as $matcher) {
+            if ($matcher->match($ip)) {
+                $trusted = true;
+                break;
+            }
+        }
+
+        $this->trustedIps->set($ip, $trusted);
+
+        return $trusted;
+    }
+
+    /**
+     * @return list<ForwardedFor>
+     */
+    private function getForwarded(Request $request): array
+    {
+        $maps = Http\parseMultipleHeaderFields($request, 'forwarded');
+        if (!$maps) {
+            return [];
+        }
+
+        $forwardedFor = [];
+        foreach ($maps as $map) {
+            $for = $map['for'] ?? null;
+            if ($for === null) {
+                continue;
+            }
+
+            $address = $this->tryInternetAddress($for);
+            if (!$address) {
+               continue;
+            }
+
+            $forwardedFor[] = new ForwardedFor($address, $map);
+        }
+
+        return $forwardedFor;
+    }
+
+    private function getForwardedFor(Request $request): array
+    {
+        $forwardedFor = Http\splitHeader($request, 'x-forwarded-for');
```

Comment:
We should test what Nginx / Apache does if the client provides a header value of `x-forwared-for: "`.

Reply:
They append the IP of the client to the list, assuming the client is itself also a proxy. Assuming that client isn't in the list of trusted proxies it doesn't really matter because we'll only use the IP appended by nginx / apache.

Reply:
No, I mean.. If the proxy just appends, quote handling will break and everything will just return null.

Reply:
If the header is invalid, the attribute set by `ForwardedForMiddleware` will be null. At that point I guess it's up to the user to decide what to do with a request that failed to parse the header.